APE 18: Adopt NEP 29 (#63)
astropy/astropy-APEs: APE18.rst

Adopt NEP 29 for CPython and Numpy Version Support
--------------------------------------------------

author: Stuart Mumford and Thomas Robitaille

date-created: 2020 October 16

date-last-revised: 2020 October 16

type: Standard Track

status: Discussion

Abstract
--------

APE 10 previously determined supported CPython versions, and provided a list up to the 4.0 release.

> Since APE 10 was written the wider scientific Python community have written and largely adopted `NEP 29`_ as a policy for supported versions of both CPython and Numpy.

Reviewer: While *I'm* fine with this, I think it might be good to give at least one concrete example of a way the more "accelerated" schedule will help things. Like what py3.7 features would we want to implement sooner rather than later? (manylinux2010 might be the answer! Or data classes?)

Author: I added a small justification in the backwards compatibility section, I don't really want to do a "let's adopt 3.7 because new features things" because I would rather this stand the test of time. I am happy to pull the justification section up here if you think it's a better place for it?

Reviewer: I see your point on "stand the test of time", but I was thinking the point of this is to give a couple of concrete examples to what is currently a very abstract phrasing.  So just meant something like "For example, at the time of the py3.7 release infrastructure maitainers found their life would be easier if manylinux2010 was an option". Or similar.

But if you really want it to be timeless, I'm ok with either pulling it up here or leaving as-is :shrug:.

This APE proposes that Astropy adopt this policy following other projects in the scientific ecosystem.

Detailed description
--------------------

This APE fully adopts `NEP 29`_ for both CPython and Numpy versions, it supersedes APE 10 wherever there is a conflict between that APE and this APE or `NEP 29`_.

While this APE will not replicate the whole of `NEP 29`_, and `NEP 29`_ should be considered the canonical source for the support policy, the core scheme to be adopted is as follows:

> The core astropy package will support:

Reviewer: One thing I'm confused by here: is this at the time of release or at the time of latest *major version* release?  That is, if we do a 4.0.x release 18 months after 4.0 is released, are we bound by the Python version from 42 months before 4.0.x or 60 months?

Reviewer 2: There are some graphics in NEP29, that makes this clear.

Reviewer 2: (except the ascii rendering is messed up, the text still makes it clear)

Author: It's based on the *scheduled* release date of non-bugfix versions, so a LTS version will support the version at the scheduled release for it's lifetime. So it would be 60 months in your example.

Reviewer: Gotcha, that definitely makes sense, but I was suggesting that we state this explicitly.  I'll make a concrete suggest for that shortly in a follow-up comment.

* All minor versions of Python released 42 months prior to the project, and at minimum the two latest minor versions.
* All minor versions of numpy released in the 24 months prior to the project, and at minimum the last three minor versions.

Backward compatibility
----------------------

This version scheme is more rapid than that proposed in APE 10, however, this allows the core library to adopt modern features, and reduces burden on maintainers.
In addition to this aligning ourselves with other projects in the ecosystem is consistent and prevents potential minimum version conflicts.
See the text of `NEP 29`_ for further justification.

> 

Reviewer: ```suggestion

This APE fully replaces APE10 where they are in conflict.

```

date-created: 2020 October 16

date-last-revised: 2020 October 16

type: Standard Track

status: Discussion

Abstract
--------

APE 10 previously determined supported CPython versions, and provided a list up to the 4.0 release.
Since APE 10 was written the wider scientific Python community have written and largely adopted `NEP 29`_ as a policy for supported versions of both CPython and Numpy.
This APE proposes that Astropy adopt this policy following other projects in the scientific ecosystem.

Detailed description
--------------------

This APE fully adopts `NEP 29`_ for both CPython and Numpy versions, it supersedes APE 10 wherever there is a conflict between that APE and this APE or `NEP 29`_.

While this APE will not replicate the whole of `NEP 29`_, and `NEP 29`_ should be considered the canonical source for the support policy, the core scheme to be adopted is as follows:

The core astropy package will support:

* All minor versions of Python released 42 months prior to the project, and at minimum the two latest minor versions.
* All minor versions of numpy released in the 24 months prior to the project, and at minimum the last three minor versions.

> 

Reviewer: ```suggestion

Note that, as stated in `NEP 29`_, these timelines all apply for the versions
set at the time of the non-bugfix release.  This means that the full
support timelines for LTS versions of astropy are 66 and 48 months for Python
and numpy, respectively.

```

Reviewer 3: Maybe worth saying since these are currently supported for two years (in case we change that some day)

Reviewer 2: I find this phrasing very confusing. 

What about rewording to say the 42 and 24 are for the start of a release "family", and support won't be dropped in a bugfix release, neither on LTS nor on normal feature release bugfix?

Reviewer 2: (also, I promised Stu a blurb about the exceptation, e.g. the latest LTS dropped a numpy version the release, as it would have been dropped  based on NEP29 very soon after the release)

Author: > What about rewording to say the 42 and 24 are for the start of a release "family", and support won't be dropped in a bugfix release, neither on LTS nor on normal feature release bugfix?

:+1: This is in NEP 29, but I am happy to call it out here in terms of the LTS if people feel it needs more clarity.

Reviewer 2: yes, I'm OK just the reference back to NEP29, it's very clear in there, but the above blurb about the 66 and 48 is very very confusing to say the least. A start of a release is a very clear date, but the end of it? And why would it even matter, by definition things are not going to be dropped between 4.0.10 and 4.0.11.

Reviewer 4: Agreed that this adds more confusion than it solves. I'd leave the text as is.

Reviewer: I disagree that it's very clear in NEP 29 for the *casual* reader.  That is it's very specific and when I read very carefully I understand the above, and really anyone in this conversation is clearly interested enough to dig deeply on this.  But I do *not* think the typical reader who comes upon this via google and just wants to know "how long will an astropy support this Python version" will take the time to click through to NEP 29, map it onto our Astropy-specific release candence, etc.

That said, if y'all think this wording is not clearer, maybe the better thing is to just give a few concrete examples and let the reader extrapolate?  Something like:
```suggestion
Note that, as stated in `NEP 29`_, these timelines all apply for the versions
set at the time of the non-bugfix release.  So for example the v4.0 release, an LTS, would support Python 3.6 (released 23 mo earlier), and would continue to support it until the end of the LTS window (24 mo after the 4.0 release).  Whereas 4.2, planned for late 2020, would  unambiguously only support 3.7 (as its full lifetime falls within the 3.7 window).
```
Or @cadirs wording from https://github.com/astropy/astropy-APEs/pull/63/files#r506678119 makes sense to me too.

Reviewer 2: See my extra sentence about the bugfix lifetimes.

Reviewer 2: (I don't find these concrete examples any useful, try to swap it out with e.g 1.3 and it will all read obsolete for the casual reader who you say would not bother to read/try to understand the tables and figures in NEP29)

Author: There is a line now stating that the versions are supported for the lifetime of the bugfixes, so hopefully that resolves the core of your comment @eteq in a way people are happy with?

Reviewer 4: Reads well to me.

Backward compatibility
----------------------

This version scheme is more rapid than that proposed in APE 10, however, this allows the core library to adopt modern features, and reduces burden on maintainers.
In addition to this aligning ourselves with other projects in the ecosystem is consistent and prevents potential minimum version conflicts.
See the text of `NEP 29`_ for further justification.

> This APE fully replaces APE10 where they are in conflict.

Reviewer 2: this should also go as an addendum for ape10

Author: sounds like a good job for the CoCo after merge ;)

Reviewer: Hmm.  I guess I'd say that *should* be part of this PR in practice, but I guess we sort of need #60 to do that :shrug:. But it would be good to make sure there's a note at the bottom that whoever merges this knows they have to do it (since it's not part of the normal merge process).

Author: Done

Adopt NEP 29 for CPython and Numpy Version Support
--------------------------------------------------

author: Stuart Mumford and Thomas Robitaille

date-created: 2020 October 16

date-last-revised: 2020 October 16

type: Standard Track

status: Discussion

Abstract
--------

APE 10 previously determined supported CPython versions, and provided a list up to the 4.0 release.
Since APE 10 was written the wider scientific Python community have written and largely adopted `NEP 29`_ as a policy for supported versions of both CPython and Numpy.
This APE proposes that Astropy adopt this policy following other projects in the scientific ecosystem.

Detailed description
--------------------

> This APE fully adopts `NEP 29`_ for both CPython and Numpy versions, it supersedes APE 10 wherever there is a conflict between that APE and this APE or `NEP 29`_.

Author: I question raised by @bsipocz is should we allow, or even mandate, that an LTS release can jump forward a numpy (or even CPython?!) version to reduce the long term maintenance burden.

I think a good way to do this would be to say something like "An LTS release of the core package may set the minimum versions for the next release" (i.e. 5.0 could adopt the versions specified by 5.1).

Reviewer 5: I would just say that institutions that are using the LTS are more likely going to have a stack that has generally older versions of Python and dependencies. So the target audience for LTS may not be happy about this.

E.g. Chandra operations is currently using Python 3.6 + numpy 1.12. We're working on an upgrade now, but it turns out to be a lot of work in our setting (review board for any software changes, etc).

Reviewer 3: i would just stick to the NEP 29 rules - if we had done for 4.0 I believe we would have required Numpy 1.15? Not too different from 1.16.

Author: The fact that it was bumped for 4.0 is the reason for this thread. That decision was clearly motivated by something. @mhvk ?

Reviewer 4: Yes, we bumped to numpy 1.16 since then we could support quantities on nearly all numpy functions, i.e., there was what was considered a very compelling reason (and LTS played a role, as with the bump the user experience would not depend on numpy version; IIRC, it also was the case that in our old scheme it would have been OK just a few months later). 

Of course, generally numpy versions do not introduce major changes that are as directly relevant to astropy, so the general rule should definitely be NEP 29. But obviously there can be some room for exceptions - I guess it would be an argued addendum to this APE perhaps, i.e., a high bar.

Reviewer 2: I would not mandate but allow. As I recall for 4.0, the timeline was just a few months early,   so it clearly made a sense both for the Quantity support, but also for planning ahead, to not to deal with 1.16 issues 2 years down the line

Reviewer 3: Yes having a clause about allowing it is very specific cases makes sense - and doesn't have to apply to just LTS eg if 4.3 came out a few days before the next bump.

Author: I have added some words. If people want to review it.

Adopt NEP 29 for CPython and Numpy Version Support
--------------------------------------------------

author: Stuart Mumford and Thomas Robitaille

date-created: 2020 October 16

date-last-revised: 2020 October 16

type: Standard Track

status: Discussion

Abstract
--------

APE 10 previously determined supported CPython versions, and provided a list up to the 4.0 release.
Since APE 10 was written the wider scientific Python community have written and largely adopted `NEP 29`_ as a policy for supported versions of both CPython and Numpy.
This APE proposes that Astropy adopt this policy following other projects in the scientific ecosystem.

Detailed description
--------------------

> This APE fully adopts `NEP 29`_ for both CPython and Numpy versions, it supersedes APE 10 wherever there is a conflict between that APE and this APE or `NEP 29`_.

Reviewer 3: ```suggestion
This APE fully adopts `NEP 29`_ for both CPython and Numpy versions for the core astropy package, and it supersedes APE 10 wherever there is a conflict between that APE and this APE or `NEP 29`_.
```

While this APE will not replicate the whole of `NEP 29`_, and `NEP 29`_ should be considered the canonical source for the support policy, the core scheme to be adopted is as follows:

The core astropy package will support:

> * All minor versions of Python released 42 months prior to the project, and at minimum the two latest minor versions.
> * All minor versions of numpy released in the 24 months prior to the project, and at minimum the last three minor versions.
> 

Reviewer 2: ```suggestion
* All minor versions of Python released 42 months prior to a non-bugfix, and at minimum the two latest minor versions.
* All minor versions of numpy released in the 24 months prior to a non-bugfix, and at minimum the last three minor versions.
* Versions of other runtime dependencies released 24 prior to a non-bugfix release.

These versions of the dependencies are kept supported for bugfix lifetime of the given astropy version.

```

Reviewer 2: So, this would introduce a policy for scipy/matplotlib/etc. We should not drop them to support only latest greatest, but then should linger with an old version neither. (inspired by @pllim seeking for input whether to drop 0.18 scipy)

Reviewer 3: I'd be up for adding such a clause, though we should make it clear this is an addition compared to NEP 29?

Reviewer 2: Yes, I'm happy to add a blurb about that if this is a go ahead.

Reviewer 4: Scipy sticks to NEP 29 itself, so we could state that in general we try to main synchronism, i.e., pick that scipy that supports all of the numpy/python we support, but not more.

Reviewer 2: I like to say we guarantee to support x versions, rather that we stay in sync, e.g. if there is no reason popping up to drop I would not drop a version (e.g., we could support very old scipies for a long time), but I would not put any extra effort in keeping up supporting an old versions when it's outside of the range.

Reviewer 4: @bsipocz - agreed there is no reason to explicitly fail on older versions (as we do for numpy); at some level it is more about what version we still test -- and for which we thus *guarantee* support.

Reviewer 2: The oldest supported one we should test with, anything older would not work as those versions are set in package metadata. I would only say we don't bump the versions unless any issues came up with them, e.g. now many versions turn out to be not python3.7 compatible, so we need to bump them along with dropping support for 3.6.

Author: I am :+1: on this suggestion

This version support policy allows, at the discretion of the maintainers, to adopt the supported version of the next release in situations where it is judged to be valuable.

> For example, in the case of an important change in numpy or a change over date very close to the release.

Reviewer 3: ```suggestion
For example, this could be done in the case of an important change in numpy or a change over date very close to the release.
```

Implementation
--------------

> Some initial work has been done in `#10900 <https://github.com/astropy/astropy/pull/10900>`__ on bumping core to Python 3.7+.

Reviewer 3: ```suggestion
Some initial work has been done in `#10900 <https://github.com/astropy/astropy/pull/10900>`__ on bumping the minimum Python version in the core astropy package to Python 3.7+.
```

Backward compatibility
----------------------

> This version scheme is more rapid than that proposed in APE 10, however, this allows the core library to adopt modern features, and reduces burden on maintainers.

Reviewer 3: ```suggestion
The version schedule here is more rapid than that proposed in APE 10, however, this allows the core library to adopt modern features, and reduces burden on maintainers.
```

> In addition to this aligning ourselves with other projects in the ecosystem is consistent and prevents potential minimum version conflicts.

Reviewer 3: ```suggestion
In addition to this, aligning ourselves with other projects in the ecosystem is consistent and prevents potential minimum version conflicts.
```

Adopt NEP 29 for CPython and Numpy Version Support
--------------------------------------------------

author: Stuart Mumford and Thomas Robitaille

date-created: 2020 October 16

date-last-revised: 2020 October 16

type: Standard Track

status: Discussion

Abstract
--------

> After the release of astropy 4.0 APE 10 no longer specifies the supported versions of CPython.

Reviewer 3: Technically APE 10 does provide a recipe for future versions. Maybe just say that APE 10 previously determined the versions of CPython we supported and provided a specific list up to 4.0 LTS or something like that?

Since APE 10 was written the wider scientific Python community have written and largely adopted `NEP 29 <https://numpy.org/neps/nep-0029-deprecation_policy.html>`__ as a policy for supported versions of both CPython and Numpy.

> This APE proposes that Astropy follow the projects who have already adopted this policy.

Reviewer 3: Adopt this policy following other projects in the scientific python ecosystem or something like that? (That is we aren't pinning our policy to other packages)

Detailed description
--------------------

This APE fully adopts NEP 29 for both CPython and Numpy versions, it supersedes APE 10 wherever there is a conflict between that APE and this APE or NEP 29.

While this APE will not replicate the whole of NEP 29, and NEP 29 should be considered the canonical source for the support policy, the core scheme to be adopted is as follows:

> astropy will support:

Reviewer 3: The core astropy package as well as other coordinated and infrastructure packages will support

Reviewer 3: This is perhaps something different from APE 10, that this would be project wide

Reviewer 2: I would not force infra and coordinated to follow this, e.g. they currently usually support more, especially the infra packages.

Reviewer 2: So I would just go that they can't drop more than what is suggested in NEP29, but they could still support older versions of they want to

Reviewer 6: Yeah, not sure how we can enforce this project-wide in a realistic manner. We barely have enough people to maintain the core as it is.

Author: I have scoped this to core, not sure we really need to do anything else here?

Reviewer 2: @pllim - and my thinking is: not sure why we _want_ to enforce this project wide.

Reviewer 2: I think it's important that none of the infra/coordinated drop things sooner and switch to python 3.10+ now, but if they have a reason to still support 3.4, then why not (usually it only means that they didn't run into an issue and everything still works, while dropping is more work)

Reviewer: Maybe we should say the "rule" applies to core, but coordinated&infrastructure packages will "aim" to follow this unless there's a reason not to?  I.e. just say for everything else it's a guideline.

I think the virtue of that is that it gives some guidance for packages that are sort of like "we don't have a rule because we haven't really thought that deeply about it".  Or alternatively/additionally, it provides an excuse to ask institutional players to help us bring things "up to standard" (I'm thinking here about how the py 3.x transition provided some people at institutions a way to fund additional work by existing Astropy devs)

Reviewer 2: many of the infra packages didn't really have a reason to drop old versions for a while, and I fail to see why it would be different in the future. As I said above, in many cases dropping a version is more work than keeping it, thus the "aim" is problematic in practice. And frankly, what is important for the ecosystem is that all coordinated and infra support at least the NEP29, that there is no situation when something gets dropped sooner than this schedule.

Reviewer 3: Agree we can make it a rule for core and recommendation for other parts of the project?

Reviewer 5: >  if they have a reason to still support 3.4, then why not 

Just thinking about #10888 that I just put in, which would require 3.7+ for astropy to work properly. If a coordinated package tries to force using Python 3.4 + astropy 4.2 then there will be problems. Not sure if that is credible.

Reviewer 2: I'm not sure what you mean credible here, it's always up to the package manager to resolve the version requirements, I don't see any force being involved here.

E.g. on 3.4, it could be possible to install the latest version of a coordinated package, and naturally, it would come with an older astropy version. It's down to the CI of the coordinated package to make sure they are compatible with their minimum supported version combination, after all the user should know best why they are stuck on an older python/numpy combo.

(3.4 is probably a bad example, but a real life example, with astroquery we had to stay python2.7 compatible much much longer than astropy core as it took long for casa to move off it. We, the astroquery maintainers decided that it's something we were willing to do, and it was totally independent of the version support policy of core that was decided separately. )

* All minor versions of Python released 42 months prior to the project, and at minimum the two latest minor versions.
* All minor versions of numpy released in the 24 months prior to the project, and at minimum the last three minor versions.

> This means that for the next few releases of astropy the following versions of CPython and Numpy will be supported:

Reviewer 3: Of astropy -> the core astropy package

+--------------------+--------------------+-----------------+--------------------+
| Astropy Release    | Release Date       | CPython Version | Numpy Version      |
+--------------------+--------------------+-----------------+--------------------+
| 4.2                | November 2020      | 3.7+            | 1.16+              |
+--------------------+--------------------+-----------------+--------------------+

> | 5.0 (LTS)          | April 2021         | 3.7+            | 1.17+              |

Reviewer 3: That doesn't seem right, 5.0 must be the following release? (4.0 was released in winter)

Reviewer 6: https://github.com/astropy/astropy/wiki/Release-Calendar doesn't say. You sure we won't have 4.3?

+--------------------+--------------------+-----------------+--------------------+
| 5.1                | November 2021      | 3.7+            | 1.18+              |
+--------------------+--------------------+-----------------+--------------------+
| 5.2                | April 2022         | 3.8+            | 1.19+              |
+--------------------+--------------------+-----------------+--------------------+
| 6.0 (LTS)          | November 2022      | 3.8+            | 1.20+              |
+--------------------+--------------------+-----------------+--------------------+
| 6.1                | April 2023         | 3.9+            | 1.20+              |
+--------------------+--------------------+-----------------+--------------------+

Backward compatibility
----------------------

> This version scheme is more rapid than that proposed in APE 10, however, this allows the core library to adopt modern features, and reduces burden on maintainers.

Reviewer 3: Also consistency with other packages

See the text of NEP 29 for further justification.

Alternatives
------------

See NEP 29.

> 

Reviewer 3: Link to NeP 29 whenever mentioned

APE 10 previously determined supported CPython versions, and provided a list up to the 4.0 release.
Since APE 10 was written the wider scientific Python community have written and largely adopted `NEP 29`_ as a policy for supported versions of both CPython and Numpy.
This APE proposes that Astropy adopt this policy following other projects in the scientific ecosystem.

Detailed description
--------------------

This APE fully adopts `NEP 29`_ for both CPython and Numpy versions for the core astropy package, and it supersedes APE 10 wherever there is a conflict between that APE and this APE or `NEP 29`_.

This APE will not replicate the content of `NEP 29`_, and the full text of `NEP 29`_ is adopted via this APE.
In addition to the content of `NEP 29`_, the APE also extends the numpy policy to all of the core package dependencies.
This leads to the following support policy for dependencies of the core package.

Astropy core will support:

* All minor versions of Python released 42 months prior to a non-bugfix, and at minimum the two latest minor versions.
* All minor versions of numpy released in the 24 months prior to a non-bugfix, and at minimum the last three minor versions.

> * Versions of other runtime dependencies released 24 prior to a non-bugfix release.

Reviewer 4: Insert months after 24

Author: ```suggestion
* Versions of other runtime dependencies released 24 months prior to a non-bugfix release.
```

These versions of the dependencies are to be supported for bugfix lifetime of the given astropy version.

This version support policy allows, at the discretion of the maintainers, to adopt the supported version of the next release in situations where it is judged to be valuable.
For example, this could be done in the case of an important change in numpy or a change over date very close to the release.

Implementation
--------------

> Some initial work has been done in `#10900 <https://github.com/astropy/astropy/pull/10900>`__ on bumping the minimum Python version in the core astropy package to Python 3.7+.
> 

Reviewer 3: The CoCo discussed this APE just now and we agree it is good to go but wanted to make a small suggestion to mention the release calendar as follows:

```suggestion
Some initial work has been done in `#10900 <https://github.com/astropy/astropy/pull/10900>`__ on bumping the minimum Python version in the core astropy package to Python 3.7+.

The release calendar for the astropy core package will be updated to include the minimum Python and Numpy version for each future release.
```

We didn't want to include the link since it's just a wiki page which may move in future. Does this seem reasonable?

Backward compatibility
----------------------

The version schedule here is more rapid than that proposed in APE 10, however, this allows the core library to adopt modern features, and reduces burden on maintainers.
In addition to this, aligning ourselves with other projects in the ecosystem is consistent and prevents potential minimum version conflicts.
See the text of `NEP 29`_ for further justification.

This APE fully replaces APE10 where they are in conflict.

Alternatives
------------

See `NEP 29`_.

Decision rationale
------------------

> <To be filled in by the coordinating committee when the APE is accepted or rejected>

Reviewer 4: I guess this line does need to be changed.
